tests/integration/utils.py

```diff
@@ -463,3 +465,166 @@ def _validate_expected_dep_file_contents(dep_contents_file: Path, output_dir: Pa
+    return path.stem
+
+
+def tested_object_name(path: Path) -> str:
+    return path.stem.lstrip("test_")
+
+
+def affects_pm(change: Path) -> bool:
+    """Check if a pm is affected.
+
+    >>> affects_pm(Path('cachi2/core/config.py'))
+    False
+    >>> affects_pm(Path('requirements.txt'))
+    False
+    >>> affects_pm(Path('tests/integration/test_gomod.py'))
+    True
+    >>> affects_pm(Path('tests/integration/utils.py'))
+    False
+    >>> affects_pm(Path('cachi2/core/package_managers/rpm/main.py'))
+    True
+    >>> affects_pm(Path('cachi2/core/package_managers/general.py'))
+    False
+    >>> affects_pm(Path('cachi2/core/package_managers/gomod.py'))
+    True
+    """
+
+    def name_belongs_to_a_pm(change: Path) -> bool:
+        return name_of(change) in supported_pms or change.parent.stem in supported_pms
+
+    def affects_pm_directly(change: Path) -> bool:
+        return "package_managers" in change.parts and name_belongs_to_a_pm(change)
+
+    def affects_pm_tests(change: Path) -> bool:
+        return tested_object_name(change) in supported_pms
+
+    return affects_pm_directly(change) or affects_pm_tests(change)
+
+
+def pm_name(pm_change: Path) -> str:
+    """Extract package manager name from a known package manager-related change.
+
+    >>> pm_name(Path('tests/integration/test_gomod.py'))
+    'gomod'
+    >>> pm_name(Path('cachi2/core/package_managers/rpm/main.py'))
+    'rpm'
+    >>> pm_name(Path('cachi2/core/package_managers/pip.py'))
+    'pip'
+    """
+    if (name := name_of(pm_change)) in supported_pms:
+        return name
+    elif (name := tested_object_name(pm_change)) in supported_pms:
+        return name
+    else:
+        return pm_change.parent.stem
+
+
+def affected_package_managers(pm_changes: tuple[Path, ...]) -> set[str]:
+    return set(pm_name(c) for c in pm_changes)
+
+
+def is_testable_code(c: Path) -> bool:
+    """Check if any actual code was affected by any of the changes.
+
+    Does this by checking if a change is in watched subtree.
+
+    >>> is_testable_code(Path('cachi2/core/config.py'))
+    True
+    >>> is_testable_code(Path('tests/integration/test_gomod.py'))
+    True
+    >>> is_testable_code(Path('tests/integration/utils.py'))
+    True
+    >>> is_testable_code(Path('tests/integration/conftest.py'))
+    True
+    >>> is_testable_code(Path('README.md'))
+    False
+    >>> is_testable_code(Path('reqruiements.txt'))
+    False
+    """
+    return any(c.is_relative_to(p) for p in paths_to_code)
+
+
+def select_testable_changes(changes: tuple[Path, ...]) -> tuple[Path, ...]:
+    """Weed out changes that cannot be tested.
+
+    If a change is outside of paths to testable code it is dropped as if
+    it never happened. Any file or module outside of watched directories
+    will be rejected.
+    """
+    return tuple(c for c in changes if is_testable_code(c))
+
+
+def just_some_package_managers_were_affected_by(changes: tuple[Path, ...]) -> bool:
```

Comment:
In context of my other comments (testing against main + affecting all pkg mgrs with changes to certain modules), we definitely DO need to test all pkg mgrs if a change was introduced to `package_managers/general.py` or essentially any non-pkg_mgr named module under the package, so this function needs to be reworked and renamed.